### PR TITLE
executor: set the correct stmtCtx for explain statement

### LINKED ISSUE #11124: actual plan of a running stmt is not the same as the result of `explain stmt`
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
``` sql
create table kankan1(id int, name text);

insert into kankan1 values(1, 'a');

insert into kankan1 values(2, 'a');

create table kankan2(id int, h1 text);

insert into kankan2 values(2, 'z');
``` 

In session1:
``` sql
tidb> explain select t1.id,sleep(20) from kankan1 t1 left join kankan2 t2 on t1.id = t2.id where (case  when t1.name='b' then 'case2' when t1.name='a' then 'case1' else NULL
+---------------------------+-------+------+-----------------------------------------------------------------------------------------+
| id                        | count | task | operator info                                                                           |
+---------------------------+-------+------+-----------------------------------------------------------------------------------------+
| Projection_6              | 1.60  | root | test.t1.id, sleep(20)                                                                   |
| └─HashLeftJoin_7          | 1.60  | root | left outer join, inner:TableReader_13, equal:[eq(test.t1.id, test.t2.id)]    -- left outer join           |
|   ├─TableReader_10        | 1.60  | root | data:Selection_9                                                                        |
|   │ └─Selection_9         | 1.60  | cop  | eq(case(eq(test.t1.name, "b"), "case2", eq(test.t1.name, "a"), "case1", NULL), "case1") |
|   │   └─TableScan_8       | 2.00  | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo                             |
|   └─TableReader_13        | 1.00  | root | data:Selection_12                                                                       |
|     └─Selection_12        | 1.00  | cop  | not(isnull(test.t2.id))                                                                 |
|       └─TableScan_11      | 1.00  | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo                             |
+---------------------------+-------+------+-----------------------------------------------------------------------------------------+
8 rows in set (0.00 sec)

tidb> select t1.id,sleep(20) from kankan1 t1 left join kankan2 t2 on t1.id = t2.id where (case  when t1.name='b' then 'case2' when t1.name='a' then 'case1' else NULL end) = 'case1';
```

In session2:
``` sql
tidb> show processlist; 
+------+------+-----------+------+---------+------+-------+------------------------------------------------------------------------------------------------------+
| Id   | User | Host      | db   | Command | Time | State | Info                                                                                                 |
+------+------+-----------+------+---------+------+-------+------------------------------------------------------------------------------------------------------+
|    1 | root | 127.0.0.1 | test | Query   |   10 | 2     | select t1.id,sleep(20) from kankan1 t1 left join kankan2 t2 on t1.id = t2.id where (case  when t1.na |
|    2 | root | 127.0.0.1 | test | Query   |    0 | 2     | show processlist                                                                                     |
+------+------+-----------+------+---------+------+-------+------------------------------------------------------------------------------------------------------+
2 rows in set (0.00 sec)

tidb> explain for connection 1;
+---------------------------+-------+------+------------------------------------------------------------------------------------------------------------------+
| id                        | count | task | operator info                                                                                                    |
+---------------------------+-------+------+------------------------------------------------------------------------------------------------------------------+
| Projection_6              | 1.25  | root | test.t1.id, sleep(20)                                                                                            |
| └─HashLeftJoin_7          | 1.25  | root | inner join, inner:TableReader_14, equal:[eq(test.t1.id, test.t2.id)]      -- inner join                                       |
|   ├─TableReader_11        | 1.60  | root | data:Selection_10                                                                                                |
|   │ └─Selection_10        | 1.60  | cop  | eq(case(eq(test.t1.name, "b"), "case2", eq(test.t1.name, "a"), "case1", NULL), "case1"), not(isnull(test.t1.id)) |
|   │   └─TableScan_9       | 2.00  | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo                                                      |
|   └─TableReader_14        | 1.00  | root | data:Selection_13                                                                                                |
|     └─Selection_13        | 1.00  | cop  | not(isnull(test.t2.id))                                                                                          |
|       └─TableScan_12      | 1.00  | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo                                                      |
+---------------------------+-------+------+------------------------------------------------------------------------------------------------------------------+
8 rows in set (0.00 sec)

```

2. What did you expect to see?
The same explain result.

3. What did you see instead?
N/A

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
`21d2590ac37bf53e5e687b8a341d50105ddbffe8` in master branch

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve? 
Fix #11124 

### What is changed and how it works?
Check the statement type of a explain statement, and set the correct stmtCtx for it.

### Check List 

Tests 

 - Integration test
TO BE ADDED.

Code changes

 - Has exported function/method change

Side effects

N/A

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
WAIT FOR #11102

**Comment by zz-jason** [comment]
@XuHuaiyu https://github.com/pingcap/tidb/pull/11105 is fixed. Please update to the master branch and resolve conflicts.

File: executor/executor.go [hunk 1]
```diff
@@ -1385,12 +1385,20 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
 		sc.MemTracker.SetActionOnExceed(action)
 	}
 
+	// execute missed stmtID uses empty sql
+	sc.OriginalSQL = s.Text()
```
Comment (qw4990) [inline executor/executor.go]: Why do we move it up to here?

**Comment by zz-jason** [comment]
@XuHuaiyu Friendly ping, any update?

File: executor/executor.go (hunk 1 above)
Comment (XuHuaiyu) [reply executor/executor.go]: Because when it's an `explainStmt`, `s` will be reset in line 1402.

**Comment by XuHuaiyu** [comment]
Ready to be reviewed. @zz-jason @qw4990 @SunRunAway @wshwsh12

File: executor/executor.go (hunk 1 above)
Comment (lamxTyler) [reply executor/executor.go]: But it will also affects the execute stmt, is it right to change it here?

Comment (XuHuaiyu) [reply executor/executor.go]: But ExecStmt is not checked in this function.

Comment (lamxTyler) [reply executor/executor.go]: For `ast.ExecuteStmt`, before this PR, original sql would be the prepared stmt like "select * from ...", but after this PR, original sql would be `execute xx`. Am I right?

Comment (XuHuaiyu) [reply executor/executor.go]: I got you.

**Pull request merged**